daft/filesystem.py

```diff
@@ -83,6 +107,161 @@ def get_filesystem_from_path(path: str, **kwargs) -> fsspec.AbstractFileSystem:
     return fs
 
 
+def _resolve_paths_and_filesystem(
+    paths: str | pathlib.Path | list[str],
+    filesystem: pa.fs.FileSystem | None = None,
+) -> tuple[list[str], pa.fs.FileSystem]:
+    """
+    Resolves and normalizes all provided paths, infers a filesystem from the
+    paths, and ensures that all paths use the same filesystem.
+
+    Args:
+        paths: A single file/directory path or a list of file/directory paths.
+            A list of paths can contain both files and directories.
+        filesystem: The filesystem implementation that should be used for
+            reading these files. If None, a filesystem will be inferred. If not
+            None, the provided filesystem will still be validated against all
+            filesystems inferred from the provided paths to ensure
+            compatibility.
+    """
+    import pyarrow as pa
+    from pyarrow.fs import (
+        FileSystem,
+        FSSpecHandler,
+        PyFileSystem,
+        _resolve_filesystem_and_path,
+    )
+
+    if isinstance(paths, pathlib.Path):
+        paths = str(paths)
+    if isinstance(paths, str):
+        paths = [paths]
+    elif not isinstance(paths, list) or any(not isinstance(p, str) for p in paths):
```

Comment:
Hmm wouldn't these be user-facing? E.g. if you passed something other than a string for a `path` arg in the read APIs, if we don't have arg type checking upstream from this, then a user could hit this, and the error message is actionable: file paths need to be strings.

But agreed that such user-facing error checking should be done at a higher level, I'll double-check that that's being done and then change these to asserts.

Reply:
Oh hmm I see... Yes in the future it would be user facing, but I guess right now only we generate the paths (by globbing) :)

In the future we can add this in the top-level API (right when they call `daft.read_*`)

Reply:
> Yes in the future it would be user facing, but I guess right now only we generate the paths (by globbing) :

Just to make sure I understand, you _can_ manually pass in a single file path right now, in addition to globbing and passing a single directory, right? I don't think that this can be hit at the `DataFrame` API level due to our automated API type checking, but I think it could be hit at the table APIs.

Reply:
Indeed - right now you can pass in:

1. Single file path
2. Glob path